[Substrait to Velox] Support the conversion between SingularOrList and Velox IN expression

rui-mo opened the pull request:
SingularOrList ([link](https://github.com/facebookincubator/velox/blob/main/velox/substrait/proto/substrait/algebra.proto#L927-L930)) in Substrait is a specialized structure that is often used is a large list of possible values. In SQL, these are typically large IN lists. For single value expressions, these are a compact equivalent of `expression = value1 OR expression = value2 OR .. OR expression = valueN`. (Reference: [or-list-equality-expression](https://substrait.io/expressions/specialized_record_expressions/#or-list-equality-expression))
This PR supports the conversion between SingularOrList and Velox IN expression. The conversion process is similar as what is done in [InTest.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/tests/InTest.cpp#L86-L90). I drew below flow chart, and hope it can help illustrate.

![Screenshot from 2023-06-06 07-11-30](https://github.com/facebookincubator/velox/assets/41687378/4389f0be-ec64-4d8d-be9b-149c3ec2a459)

In this picture, `value` and `options` are two items of SingularOrList. The value is the variable, and the options are the possible values to be compared with. They represent two arguments for the creation of a Velox IN expression.

zhejiangxiaomai commented on velox/substrait/VeloxToSubstraitExpr.h:
```diff
@@ -96,6 +96,13 @@ class VeloxToSubstraitExprConvertor {
+  /// Convert Velox IN expression to Substrait SingularOrList.
+  const ::substrait::Expression& toSubstraitSingularOrList(
+      google::protobuf::Arena& arena,
+      const std::vector<std::shared_ptr<const core::ITypedExpr>>& inputs,
```
maybe use TypedExprPtr to replace std::shared_ptr<const core::ITypedExpr>.

rui-mo replied:
Fixed cases like this. Thanks for your review.

zhejiangxiaomai commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,6 +340,46 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+    literals.emplace_back(option.literal());
+  }
+
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
```
maybe use  std::vector<TypedExprPtr> params;

zhejiangxiaomai commented on velox/substrait/VeloxToSubstraitExpr.cpp:
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
+const ::substrait::Expression&
+VeloxToSubstraitExprConvertor::toSubstraitSingularOrList(
+    google::protobuf::Arena& arena,
+    const std::vector<std::shared_ptr<const core::ITypedExpr>>& inputs,
```
const std::vector<TypedExprPtr>& inputs

rui-mo commented:
Hi @mbasmanova, would you like to review this PR? Thanks.

rui-mo commented:
hi @mbasmanova @vibhatha, could you help review this PR when free? Thanks.

rui-mo commented:
Hi @mbasmanova, would you like to review this PR? Or anyone else can help on that? Thanks.

mbasmanova commented:
@rui-mo Rui, can you verify that the semantics of these two are exactly the same and explain how you did that in the PR description? Also, CI is failing. Please, take a look.

rui-mo commented:
> @rui-mo Rui, can you verify that the semantics of these two are exactly the same and explain how you did that in the PR description? Also, CI is failing. Please, take a look.

Hi @mbasmanova, I updated the PR description to explain the semantics and the conversion process. CI was also fixed. Could you check again? Thank you.

mbasmanova commented:
@rui-mo Rui, this description is very helpful. Thank you for putting it together. One question: why does the diagram has 2 paths? How do you decide which path to take?

mbasmanova commented on velox/substrait/VeloxToSubstraitExpr.cpp:
```diff
@@ -462,6 +495,10 @@ const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
   auto inputs = callTypeExpr->inputs();
   auto& functionName = callTypeExpr->name();
 
+  if (functionName == "in") {
```
Hardcoding the name of the function is problematic. This function may be registered under a different name or a different function can be registered under the "in" name.

I wonder if a more robust approach is to introduce Options struct to specify various configurations for the converter and have one of these options be the name of the "IN" function.

rui-mo replied:
Understand the issue here. I'm having two questions here:

- How do we know the exact `callTypeExpr->name()` of IN expression in Velox? It could vary according to the prefix.
- Can we record the prefix during registration and append it here?

mbasmanova replied:
> How do we know the exact callTypeExpr->name() of IN expression in Velox?

We don't. Hence, I suggest we require that the user of the converter specify it.

rui-mo replied:
I added prefix in the converter, because this is not just an issue for `IN`. The prefix of other scalar functions also needs to be considered.

mbasmanova replied:
@rui-mo Rui, I'll take a look, but I don't think it is sufficient to specify he prefix. I imagine we need to specify full function names because it is not guaranteed that "in" name refers to the "right" function.

rui-mo replied:
I find presto and spark both register it as `in`, seeing [GeneralFunctionsRegistration.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/registration/GeneralFunctionsRegistration.cpp#L24) and [In.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/In.cpp#L117). How about using certain definition just like aggregate names in [AggregateNames.h](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/AggregateNames.h)?

mbasmanova commented on velox/substrait/VeloxToSubstraitExpr.cpp:
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
+    ::substrait::Expression* substraitExpr) {
+  ::substrait::Expression_SingularOrList* singularOrList =
+      substraitExpr->mutable_singular_or_list();
+  VELOX_CHECK(inputs.size() == 2, "Two arguments expected for IN.");
```
Use VELOX_CHECK_EQ for better error messages.

rui-mo replied:
Fixed.

mbasmanova commented on velox/substrait/VeloxToSubstraitExpr.cpp:
```diff
@@ -452,6 +452,39 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
+  VELOX_CHECK_NOT_NULL(constantExpr);
+  VELOX_CHECK(
+      constantExpr->hasValueVector(),
+      "Value vector expected for constant expression.");
```
Should we also check the type of the expression? Should it be ARRAY(BIGINT())?

rui-mo replied:
Added type check for ARRAY, but the element type can be of other type except for BIGINT.

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -40,12 +40,12 @@ class SubstraitVeloxExprConverter {
 
   /// Convert Substrait ScalarFunction into Velox Expression.
-  std::shared_ptr<const core::ITypedExpr> toVeloxExpr(
+  core::TypedExprPtr toVeloxExpr(
```
Thank you for refactor. It would be nice to submit it as a separate PR to keep changes in this PR small.

rui-mo replied:
Sure. Submitted https://github.com/facebookincubator/velox/pull/5177.

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
     const ::substrait::Expression& substraitExpr,
     const RowTypePtr& inputType) {
-  std::shared_ptr<const core::ITypedExpr> veloxExpr;
+  core::TypedExprPtr veloxExpr;
```
Is this variable used? If not, let's remove it.

rui-mo replied:
Fixed.

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+core::TypedExprPtr SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::SingularOrList& singularOrList,
+    const RowTypePtr& inputType) {
+  VELOX_CHECK(
```
VELOX_CHECK_GT

rui-mo replied:
Fixed.

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+  params.emplace_back(toVeloxExpr(singularOrList.value(), inputType));
+  params.emplace_back(literalsToConstantExpr(literals));
+  return std::make_shared<const core::CallTypedExpr>(
+      BOOLEAN(), std::move(params), "in");
```
Same comment about hard-coding "in".

rui-mo replied:
Fixed.

rui-mo commented:
> @rui-mo Rui, this description is very helpful. Thank you for putting it together. One question: why does the diagram has 2 paths? How do you decide which path to take?

@mbasmanova Sorry for the misleading. They are not two paths, but two items of  below `SingularOrList`.  The value is the variable, and the options are the possible values to be compared with. They are two arguments for the creation of a Velox IN expression.

```
  message SingularOrList {
    Expression value = 1;
    repeated Expression options = 2;
  }
```

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+  VELOX_CHECK(
+      singularOrList.options_size() > 0, "At least one option is expected.");
+  auto options = singularOrList.options();
+  std::vector<::substrait::Expression::Literal> literals;
```
Can we avoid creating this vector and create a vector of TypedExprPtr instead?

rui-mo replied:
Fixed.

mbasmanova commented on velox/substrait/tests/VeloxSubstraitRoundTripTest.cpp:
```diff
@@ -402,6 +402,16 @@ TEST_F(VeloxSubstraitRoundTripTest, topNTwoKeys) {
+  createDuckDbTable(vectors);
+  auto plan = PlanBuilder(pool_.get())
+                  .values(vectors)
+                  .filter("c0 in (1, 2, 3)")
```
Let's add tests for all supported types of IN values.

mbasmanova replied:
> Current code goes to prestosql, and more types are supported in the IN function under sparksql. Do we need to use that?

It would be nice to test with both.

mbasmanova replied:
> For bigint and string, an error in duckdb parser is found at [DuckParser.cpp](https://github.com/facebookincubator/velox/blob/main/velox/duckdb/conversion/DuckParser.cpp#L373). The reason is the parser adds some casting to generate a literal of certain type.

Maybe this can be fixed (in a separate PR).

rui-mo replied:
> It would be nice to test with both.

Tests were added for IN of presto and spark sql function.

> Maybe this can be fixed (in a separate PR).

Added TODO in this PR. More debugging is needed for the conversion from string to expression lists in DuckDB parser.

mbasmanova replied:
> Added TODO in this PR. More debugging is needed for the conversion from string to expression lists in DuckDB parser.

Sounds good. Would you file an issue to explain what the problem is and maybe provide a small repro?

mbasmanova commented:
> They are not two paths, but two items of below `SingularOrList`. The value is the variable, and the options are the possible values to be compared with. They are two arguments for the creation of a Velox IN expression.

@rui-mo Got it. Thank you for clarifying. Perhaps, update the PR description for the benefit of future readers.

rui-mo commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -339,11 +338,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
+  }
+  auto varArray = variant::array(variants);
+  ArrayVectorPtr arrayVector =
+      core::variantArrayToVector(varArray.inferType(), varArray.array(), pool_);
```
Removed the use of inferType(), and use element type instead.

rui-mo commented:
Hi @mbasmanova, thanks for your review. I fixed most comments, and have some questions on the name and unit test. Could you take a look again? Thank you.

rui-mo commented:
Hi @mbasmanova, I fixed the hard-coded function name, and enabled spark `IN` tests. Could you review again? Thanks.

mbasmanova commented on velox/duckdb/conversion/DuckParser.h:
```diff
@@ -30,6 +30,9 @@ struct ParseOptions {
   bool parseIntegerAsBigint = true;
+
+  // The prefix of registered functions.
+  std::string functionPrefix = "";
```
I'm not sure I understand this change. Would you clarify. Changes to DuckParser need to be submitted in a separate PR with a test.

rui-mo replied:
Previously, DuckParser generates `in` by default which goes into presto sql function, when spark sql functions were registered with prefix `spark`. Here, I added the prefix as an option to let the user can control the generated function.

rui-mo replied:
Opened https://github.com/facebookincubator/velox/pull/5212 for this change. If the idea is OK, I can apply function prefix on all the needed places in DuckParser.

rui-mo commented on velox/duckdb/conversion/DuckParser.cpp:
```diff
@@ -378,7 +378,8 @@ std::shared_ptr<const core::IExpr> parseOperatorExpr(
         ARRAY(valueType), variant::array(values), std::nullopt));
-    auto inExpr = callExpr("in", std::move(params), getAlias(expr));
+    auto inExpr = callExpr(
+        options.functionPrefix + "in", std::move(params), getAlias(expr));
```
Prefix was also applied to other scalar functions.
https://github.com/facebookincubator/velox/pull/4690/files#diff-9a83c319e87e7171bc95fdd1ccb3f2af4c19e6f3a7048c095f4238a955a74d1dR228

rui-mo replied:
Yes, only `in` was changed to ensure this PR is small. Maybe other functions' change should be put in another PR if needed. How do you think?

mbasmanova commented on velox/substrait/VeloxToSubstraitPlan.h:
```diff
@@ -36,11 +36,13 @@ class VeloxToSubstraitPlanConvertor {
   /// Convert Velox PlanNode into Substrait Plan.
   /// @param vPlan Velox query plan to convert.
   /// @param arena Arena to use for allocating Substrait plan objects.
+  /// @param prefix Registered function prefix.
```
Would you clarify what it is used for / applied to?

rui-mo replied:
Added more info for it.

mbasmanova commented on velox/substrait/SubstraitToVeloxPlan.h:
```diff
@@ -80,7 +80,10 @@ class SubstraitVeloxPlanConverter {
+  /// Prefix: the prefix of registered functions.
+  core::PlanNodePtr toVeloxPlan(
+      const ::substrait::Plan& substraitPlan,
+      const std::string& prefix = "");
```
Shouldn't the prefix be passed to the constructor?

rui-mo replied:
Moved to the constructor.

mbasmanova commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -63,6 +64,22 @@ class SubstraitVeloxExprConverter {
+          constantExprs);
+
+  /// Return the function prefix.
+  std::string getFuncPrefix() {
```
use const& for return value and make this method const

rui-mo replied:
Fixed.

mbasmanova reviewed (commented):
@rui-mo Looks good to me % a few comments.

rui-mo commented:
> @rui-mo Looks good to me % a few comments.

@mbasmanova Fixed above comments. Could you help review again? Thank you.

mbasmanova commented on velox/substrait/SubstraitParser.h:
```diff
@@ -72,9 +72,11 @@ class SubstraitParser {
       const std::unordered_map<uint64_t, std::string>& functionMap,
-      uint64_t id) const;
+      uint64_t id,
+      const std::string& prefix = "") const;
```
Do we need a default value here? If not, let's remove.

rui-mo replied:
Removed.

mbasmanova commented on velox/substrait/SubstraitParser.h:
```diff
@@ -72,9 +72,11 @@ class SubstraitParser {
 
   /// Find the Velox function name according to the function id
   /// from a pre-constructed function map.
+  /// prefix: the prefix of registered SQL functions.
```
@param prefix ...

rui-mo replied:
Fixed.

mbasmanova reviewed (approved):
@rui-mo Looks good to me.

FelixYBW commented:
@rui-mo  let's close this one to clear up the pending PRs
